### PR TITLE
release: promote Backend dev → main for Sprint 6 S1 [KAN-213]

### PR DESCRIPTION
**Pull request by adamtasteslikegood**
## Summary

Promotes Backend `dev` → `main` for the Sprint 6 release. Primary change:

- **KAN-213**: Server-side duplicate-recipe refusal via two partial unique indexes on `COALESCE(source_slug, slug)`. Concurrent saves of the same recipe return 409 instead of silently creating a duplicate. 536-line test suite covering the race, guest paths, update collisions, and edge cases.

Also includes docs and CI dependency bumps merged since the last promotion.

## Test plan

- [x] Backend CI all green on dev (PR #273 + subsequent PRs)
- [ ] Merge this promotion, then back-sync main → dev

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01WoJz3wTTx2R9oxWFTN8PBN

---
Rovo Dev code review: <strong>Out of Rovo Dev credits</strong>
You've used all your Rovo Dev credits, so Rovo Dev can't review your pull requests.

### CODE REVIEW

File: repositories/db_recipe_repository.py [hunk 1]
```diff
@@ -257,11 +291,80 @@ def _resolve_public_slug(
...
+
+    Scoped to the acting owner on purpose: two different people saving the same
+    public recipe is the product working, not a duplicate, and the partial
+    indexes are per-owner for that reason.
+    """
+    if owner_scope is None:
+        return False
+    # Mirrors the index key COALESCE(source_slug, slug): a row's identity is the
+    # public recipe it points at, or its own page when it is the original.
+    identity = recipe_data.get("sourceSlug") or recipe_data.get("slug")
```
[the remark this answers is not in the document]
Comment (adamtasteslikegood) [reply repositories/db_recipe_repository.py]: Confirmed — independently reproduced via an out-of-band `/codex` CLI review (separate from the in-repo automated review action), triangulating with this comment.

`recipe_data.get("sourceSlug") or recipe_data.get("slug")` diverges from `func.coalesce(Recipe.source_slug, Recipe.slug)` on exactly the case you named: `sourceSlug: ""` is falsy in Python but non-NULL in SQL, so this detector misses a write the index actually refused, and it falls through to the raw/generic integrity path (500) instead of `RecipeDuplicateError` (409).

Reachability today: the SPA never sends an empty string (`public-recipe.mapper.ts` only sets a real slug or omits the key), and `recipe_schema.json` doesn't declare `sourceSlug` at all, so nothing currently validates it — any direct API caller can hit this. Not a live user-facing bug, but a real latent one.

Fix: `identity = recipe_data.get("sourceSlug"); identity = identity if identity is not None else recipe_data.get("slug")` (or a small helper shared with the SQL side) — explicit `is not None`, not truthiness.

_Replied by Claude on Adam's behalf_

File: blueprints/auth_api_bp.py [hunk 2]
```diff
@@ -144,6 +144,46 @@ def _merge_guest_session_into_user(user, guest_session_id, max_retries=3):
...
+                    # `recipe.source_slug = data.get('sourceSlug')`. Clearing
+                    # only the column means the next ordinary partial PUT (say
+                    # {"name": "..."}) pulls the stale value back out of the
+                    # untouched blob and writes it to the column — resurrecting
+                    # the duplicate and making the row un-editable behind a 409.
+                    # Caught by Codex review on PR #273; reproduced by
+                    # test_clearing_the_column_survives_a_later_partial_update.
+                    recipe.source_slug = None
+                    if recipe.data and "sourceSlug" in recipe.data:
+                        del recipe.data["sourceSlug"]
```
[the remark this answers is not in the document]
Comment (adamtasteslikegood) [reply blueprints/auth_api_bp.py:190]: Confirmed as a real gap in the logic as written — independently reproduced via `/codex` CLI. `_recipe_identity_keys()` matches on `{source_slug, slug}`, but the clear only touches `source_slug`. If the match came from this row's own `slug` equaling an owned row's `COALESCE(source_slug, slug)`, the post-clear identity (now falling back to this row's `slug`) is exactly the value that collided in the first place — the clear doesn't remove the row from the index's coverage in that sub-case, and `IntegrityError` repeats until retries exhaust.

That said: per the comment three lines above this one (and confirmed separately in a prior session's notes on KAN-213), the branch this sits in is currently dead — `_gate_is_public()` forces `is_public=False` for every guest write, so no guest row can be `is_public=True` today, and migration `e91b47a2c5d3` already swept legacy rows that predated the gate. So this isn't reachable against current data or current code paths.

Flagging anyway because it's exactly the shape of trap worth closing while cheap: if the `is_public` gate is ever weakened by a future change, this code reawakens with precisely the failure mode you describe — the entire login merge rolled back, guest recipes and cookbooks orphaned. Recommend resolving the post-clear identity explicitly (verify no collision remains, or remap rather than merely clearing) rather than leaving it as a guard that only works for one of the two ways a match can occur.

_Replied by Claude on Adam's behalf_

File: tests/test_source_slug_unique.py [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+cross-context race by construction: two tabs, or a tab and a phone, both read
+"you don't have this yet" before either writes.
+
+These tests pin the fix at the only layer that can refuse the second write —
+two partial unique indexes:
+
+    uq_recipe_user_source_slug   UNIQUE (user_id, source_slug)
+        WHERE source_slug IS NOT NULL AND user_id IS NOT NULL
+    uq_recipe_guest_source_slug  UNIQUE (guest_session_id, source_slug)
+        WHERE source_slug IS NOT NULL AND guest_session_id IS NOT NULL
```
[the remark this answers is not in the document]
Comment (adamtasteslikegood) [reply tests/test_source_slug_unique.py]: Confirmed, agreed — the module docstring should describe `COALESCE(source_slug, slug)`, not source_slug alone, since the own-published-recipe-collision case this file tests for only exists because of the COALESCE fallback.

_Replied by Claude on Adam's behalf_

File: models/recipe.py [hunk 4]
```diff
@@ -1,12 +1,68 @@
...
+    # KAN-213 — the duplicate invariant lives here, not in the SPA.
+    #
+    # Exactly one of (user_id, guest_session_id) is non-NULL on a row, so two
+    # partial indexes are the right shape rather than one composite constraint
+    # — the same reasoning as cookbook's uq_cookbook_* pair.
+    #
+    # Partial on `source_slug IS NOT NULL` by design, and narrower than that
+    # reads: only `origin='saved'` rows carry a source_slug, so these indexes
+    # constrain COPIES A USER TOOK from someone else's public page and do not
+    # constrain a single recipe a user authored (~3% of user 1's rows).
```
[the remark this answers is not in the document]
Comment (adamtasteslikegood) [reply models/recipe.py]: Confirmed — independently reproduced via `/codex` CLI, which flagged the identical mismatch. The comment states the partial predicate as `source_slug IS NOT NULL` and derives "~3% of rows" / "does not constrain a single recipe a user authored" from that, but `_IDENTITY = coalesce(source_slug, slug)` and the predicate is `{_IDENTITY} IS NOT NULL` — which nearly every row satisfies (slug is populated on essentially all non-legacy rows). The behavior is right (this broader scope is exactly what closes the own-published-recipe hole described a few lines below in the same comment block) — the description of *what SQL predicate governs inclusion* is what's stale/self-contradictory. Worth tightening so a future reader doesn't reason from the "~3%" framing the way this comment currently invites.

_Replied by Claude on Adam's behalf_

File: migrations/versions/c8f3b71d20a4_recipe_source_slug_unique_indexes.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    bind = op.get_bind()
+
+    # Hold writers out of the gap between the de-dup pass and the index build.
+    # Held until this migration's transaction commits, i.e. past index creation.
+    # Postgres only: SQLite has no LOCK TABLE and runs single-connection anyway.
+    if bind.dialect.name == "postgresql":
+        op.execute("LOCK TABLE recipe IN ACCESS EXCLUSIVE MODE")
+
+    cleared = _clear_duplicate_identities(bind, "user_id")
+    cleared += _clear_duplicate_identities(bind, "guest_session_id")
```
[the remark this answers is not in the document]
Comment (adamtasteslikegood) [reply migrations/versions/c8f3b71d20a4_recipe_source_slug_unique_indexes.py]: Confirmed as a real gap — independently reproduced via `/codex` CLI with the same shape of counter-example: user-scoped rows A (`source_slug='orig'`), B (`source_slug='orig', slug='copy'`), C (`source_slug='copy'`). The single pass sees B's group as `{A, B}` (identity `orig`, from the snapshot taken *before* any clears), keeps A, clears B. But B's *new* identity (post-clear, `COALESCE(NULL, 'copy')` = `copy`) now collides with C, which the snapshot never re-evaluates. `op.create_index(..., unique=True)` then fails inside the same transaction (LOCK TABLE is still held), aborting the whole migration — fail-closed per the runbook, not data corruption, but it blocks `flask-backend-migrate` and the deploy.

Currently low-risk: the docstring notes `cleared` is expected to be 0 in production (purged by hand, gates verified empty), so this chain shape shouldn't exist in the data this migration will actually run against. But the dedup pass as written doesn't converge to a fixed point, which matters the moment this pattern gets reused or the assumption about clean production data doesn't hold on a future run.

Fix: loop `_clear_duplicate_identities` per scope until it returns 0 in the same call (re-querying each iteration), before creating the indexes — rather than the current single pass per scope.

_Replied by Claude on Adam's behalf_

File: migrations/versions/c8f3b71d20a4_recipe_source_slug_unique_indexes.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+reasoning as the cookbook pair in b7e2a9c4d1f8. Both ship together or neither
+(Sprint 6 R3): guests key on guest_session_id, which is the KAN-186 path.
+
+Partial on ``source_slug IS NOT NULL`` deliberately, and the scope is narrower
+than "most of the table is excluded" suggests. Only ``origin = 'saved'`` rows
+carry a source_slug (the SPA sets origin and sourceSlug together when saving
+from a public page), so:
+
+    These indexes constrain only copies a user took from someone else's public
+    page. They do not constrain a single recipe a user authored.
```
[the remark this answers is not in the document]
Comment (adamtasteslikegood) [reply migrations/versions/c8f3b71d20a4_recipe_source_slug_unique_indexes.py]: Same issue as the `models/recipe.py` comment (also flagged there) — confirmed. The predicate is `COALESCE(source_slug, slug) IS NOT NULL`, not `source_slug IS NOT NULL`; the "roughly 3% coverage" framing describes only the `source_slug`-populated subset and understates what the partial index actually includes. Worth correcting since this docstring is the reference future operators will read when judging this migration's blast radius before a production run.

_Replied by Claude on Adam's behalf_

**Review by adamtasteslikegood — COMMENTED** [review]
## Independent `/codex` review

This is a fresh-eyes pass on top of the reviews already on this PR (the automated "Independent Claude review" action, which found the two-invariant-hole fixes correct and said "no changes needed"; and Copilot's 6 inline comments, none yet replied to). Ran the OpenAI Codex CLI against the full diff with an adversarial-review prompt (not shown Copilot's comments beforehand), then manually verified every claim against `origin/dev` before posting anything.

**Result: Codex independently converged on the same 3 substantive findings Copilot already left inline** — same functions, same failure modes, one with the identical counter-example. Replied to each of the 6 Copilot threads individually with verification detail; summary here:

| Finding | File | Verdict | Currently reachable? |
|---|---|---|---|
| `or` vs `COALESCE` on empty string | `db_recipe_repository.py:343` | **Confirmed bug** | Not via SPA (never sends `""`), but `recipe_schema.json` doesn't validate `sourceSlug` at all — any direct API caller can hit it |
| Guest-merge clear misses slug-side match | `auth_api_bp.py:185` | **Confirmed bug as written** | No — sits in a branch already documented (and confirmed in a prior KAN-213 session) as dead: guests can't publish, legacy rows were swept by migration `e91b47a2c5d3` |
| Migration dedup is not a fixed point | `c8f3b71d20a4_...py:193` | **Confirmed bug**, reproduced with the same 3-row example independently by Codex | Low — docstring notes `cleared` is expected 0 in production (purged + gated), but the pass doesn't converge if that ever isn't true |
| Doc says `source_slug IS NOT NULL`, predicate is `COALESCE(...)  IS NOT NULL` | `models/recipe.py:20`, migration `:32` | **Confirmed**, self-contradicts a later paragraph in the same comment block | N/A — documentation only |
| Stale module docstring | `test_source_slug_unique.py:17` | Confirmed, minor | N/A — documentation only |

None of these are live user-facing bugs right now — each is gated by an invariant this codebase has independently earned (schema validation gap aside, empty-string never sent; guest-publish gate; expected-clean production data). That's why the automated "no changes needed" review wasn't wrong about user impact. But three of them are real correctness gaps that will fire silently the moment one of those invariants stops holding, and two are one-line fixes.

**Recommendation:** Fix the `or`→`is not None` bug (`db_recipe_repository.py:343`) and the two doc-scope corrections before merging — both are trivial and this migration/index pair is exactly the kind of code future changes will read the comments to reason about. The migration fixed-point bug and the guest-merge gap are lower urgency (both currently unreachable) and reasonable to track as fast-follows rather than block this dev→main promotion on, given CI is green and production data is expected clean — but they shouldn't go unaddressed a second time, because this is the same code that already went through three rounds of exactly this kind of review on #273 (see [the Sprint 6 retro](https://github.com/adamtasteslikegood/tasteslikegoodtheangularsvegancookbook/discussions/3376), which is directly about this pattern). Commenting there with this as a fourth data point.

Recommendation: fix the two one-line issues before merge because they're cheap and this file is the reference other changes will reason from; track the other two as fast-follows because they're currently unreachable given the SPA/gate invariants documented in-repo, and CI is otherwise green.

_Replied by Claude on Adam's behalf_
